### PR TITLE
cabal-install: Fix non-reinstallable package set

### LINKED ISSUE #9064: ghc-boot is reinstallable
~~#8489~~ #8501 added `ghc-boot`  to the `nonUpgradeablePackages` list but I don't see why `ghc-boot` is not reinstallable. It is a normal dependency of `ghc` like `containers` (which is reinstallable). 

It would be good to document what a `nonUpgradeablePackage` is to avoid additional packages being added here in future.

### PR DESCRIPTION
**Pull request by bgamari**
In #9064 we discovered that `ghc-boot` was added to the non-reinstallable package set due to #8501 despite there being no reason why it can't be built from its source distribution. This revealed the fact that there is quite some ambiguity around what constitutes a non-reinstallable package.

In #9064 we worked out a hopefully-more-clear picture of non-reinstallability. Here we update the commentary to describe this concept and update the lists to reflect the new definition.

Closes #9064.

### CODE REVIEW

**Review by ulysses4ever — APPROVED** [review]
Thank you!

**Review by geekosaur — APPROVED** [review]

**Review by gbaz — APPROVED** [review]

File: cabal-install/src/Distribution/Client/Dependency.hs [hunk 1]
```diff
@@ -456,19 +456,34 @@ dontUpgradeNonUpgradeablePackages params =
...
 nonUpgradeablePackages :: [PackageName]
 nonUpgradeablePackages =
   [ mkPackageName "base"
   , mkPackageName "ghc-bignum"
   , mkPackageName "ghc-prim"
-  , mkPackageName "ghc-boot"
   , mkPackageName "ghc"
-  , mkPackageName "ghci"
   , mkPackageName "integer-gmp"
   , mkPackageName "integer-simple"
```
Comment (hsyl20) [inline cabal-install/src/Distribution/Client/Dependency.hs:488]: Both `integer-*` packages are reinstallable now afaik.

**Comment by ulysses4ever** [comment]
@bgamari do you need help with this?

**Review by grayjay — APPROVED** [review]
Thank you for documenting this!

File: cabal-install/src/Distribution/Client/Dependency.hs (hunk 1 above)
Comment (Kleidukos) [reply cabal-install/src/Distribution/Client/Dependency.hs:488]: @hsyl20 Is there an authoritative source about this? I can ship this change in the PR if you have a source :)

Comment (hsyl20) [reply cabal-install/src/Distribution/Client/Dependency.hs:488]: No authoritative source, sorry.

Comment (gbaz) [reply cabal-install/src/Distribution/Client/Dependency.hs:488]: Even if this were true, that's only for latest ghcs, and this package set needs to span all supported ghcs, so we can't do it regardless. If we do start doing it going forward we'll need a per-ghc-version refactor of how we manage the set.

**Comment by gbaz** [comment]
Is this blocked or can we merge?

**Comment by ulysses4ever** [comment]
@gbaz I think it has been unblocked since I applied the label, so I'm removing the label now.

If no one feels too strong about `integer-simple`, we can put the `squash+merge` label. This will still leave 2 days for contemplating, and the timer will reset if someone decides to add more comments. 

Perhaps we need a ticket for making the list of non-upgradable packages dependent on the compiler version?

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs [hunk 2]
```diff
@@ -1301,8 +1295,6 @@ dbBase =
 dbNonupgrade :: ExampleDb
 dbNonupgrade =
   [ Left $ exInst "ghc" 1 "ghc-1" []
-  , Left $ exInst "ghci" 1 "ghci-1" []
-  , Left $ exInst "ghc-boot" 1 "ghc-boot-1" []
   , Right $ exAv "ghc" 2 []
   , Right $ exAv "ghci" 2 []
   , Right $ exAv "ghc-boot" 2 []
```
Comment (grayjay) [inline cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs:1300]: The source versions of ghci and ghc-boot and packages B and C should also me removed.

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.10

**Comment by Kleidukos** [comment]
@grayjay damnit, sorry I missed your message :/

**Pull request merged**